branch.sql

```diff
@@ -0,0 +1,54 @@
+BEGIN;
+
+DELETE FROM communities WHERE nmembers = 0;
+ALTER TABLE communities DROP CONSTRAINT communities_nmembers_check,
+                        ADD CONSTRAINT communities_nmembers_check CHECK (nmembers > 0);
+
+CREATE OR REPLACE FUNCTION upsert_community() RETURNS trigger AS $$
+    DECLARE
+        is_member boolean;
+        delta int = CASE WHEN NEW.is_member THEN 1 ELSE -1 END;
+    BEGIN
+        IF (SELECT is_suspicious FROM participants WHERE id = NEW.participant) THEN
+            RETURN NULL;
+        END IF;
+        is_member := (
+            SELECT cur.is_member
+              FROM community_members cur
+             WHERE slug = NEW.slug
+               AND participant = NEW.participant
+          ORDER BY mtime DESC
+             LIMIT 1
+        );
+        IF (is_member IS NULL AND NEW.is_member IS false OR NEW.is_member = is_member) THEN
+            RETURN NULL;
+        END IF;
+        LOOP
+            UPDATE communities
+               SET nmembers = nmembers + delta
+             WHERE slug = NEW.slug
+               AND nmembers + delta > 0;
+            EXIT WHEN FOUND;
+            IF (NEW.is_member) THEN
+                BEGIN
+                    INSERT INTO communities
+                         VALUES (NEW.slug, NEW.name, 1, NEW.ctime);
+                EXCEPTION
+                    WHEN unique_violation THEN
+                        IF (CONSTRAINT_NAME = 'communities_slug_pkey') THEN
+                            CONTINUE; -- Try again
+                        ELSE
+                            RAISE;
+                        END IF;
+                END;
+                EXIT;
+            ELSE
+                DELETE FROM communities WHERE slug = NEW.slug AND nmembers = 1;
```

Comment:
`communities` is an INSERT-only table. Why should we start DELETEing from it here?